pkg/builder/build_test.go

```diff
@@ -119,94 +123,157 @@ var _ = Describe("application", func() {
-			})
-
-			instance, err := SimpleController().ForType(&appsv1.Deployment{}).
+	Describe("Start with SimpleController", func() {
+		It("should Reconcile Owns objects", func(done Done) {
+			bldr := SimpleController().
+				ForType(&appsv1.Deployment{}).
 				WithConfig(cfg).
-				Owns(&appsv1.ReplicaSet{}).
-				Build(fn)
+				Owns(&appsv1.ReplicaSet{})
+			doReconcileTest("1", stop, bldr, nil, false)
+
+			close(done)
+		}, 10)
+
+		It("should Reconcile Owns objects with a Manager", func(done Done) {
+			m, err := manager.New(cfg, manager.Options{})
 			Expect(err).NotTo(HaveOccurred())
 
-			By("Starting the application")
-			go func() {
-				defer GinkgoRecover()
-				Expect(instance.Start(stop)).NotTo(HaveOccurred())
-				By("Stopping the application")
-			}()
-
-			By("Creating a Deployment")
-			// Expect a Reconcile when the Deployment is managedObjects.
-			dep := &appsv1.Deployment{
-				ObjectMeta: metav1.ObjectMeta{
-					Namespace: "default",
-					Name:      "deploy-name",
-				},
-				Spec: appsv1.DeploymentSpec{
-					Selector: &metav1.LabelSelector{
-						MatchLabels: map[string]string{"foo": "bar"},
-					},
-					Template: corev1.PodTemplateSpec{
-						ObjectMeta: metav1.ObjectMeta{Labels: map[string]string{"foo": "bar"}},
-						Spec: corev1.PodSpec{
-							Containers: []corev1.Container{
-								{
-									Name:  "nginx",
-									Image: "nginx",
-								},
-							},
-						},
-					},
-				},
-			}
-			err = instance.GetClient().Create(context.TODO(), dep)
+			bldr := SimpleController().
+				WithManager(m).
+				For(&appsv1.Deployment{}).
+				Owns(&appsv1.ReplicaSet{})
+			doReconcileTest("2", stop, bldr, m, false)
+			close(done)
+		}, 10)
+	})
+
+	Describe("Start with ControllerManagedBy", func() {
+		It("should Reconcile Owns objects", func(done Done) {
+			m, err := manager.New(cfg, manager.Options{})
+			Expect(err).NotTo(HaveOccurred())
+
+			bldr := ControllerManagedBy(m).
+				For(&appsv1.Deployment{}).
+				Owns(&appsv1.ReplicaSet{})
+			doReconcileTest("3", stop, bldr, m, false)
+			close(done)
+		}, 10)
+
+		It("should Reconcile Watches objects", func(done Done) {
+			m, err := manager.New(cfg, manager.Options{})
 			Expect(err).NotTo(HaveOccurred())
 
-			By("Waiting for the Deployment Reconcile")
-			Expect(<-ch).To(Equal(reconcile.Request{
-				NamespacedName: types.NamespacedName{Namespace: "default", Name: "deploy-name"}}))
-
-			By("Creating a ReplicaSet")
-			// Expect a Reconcile when an Owned object is managedObjects.
-			t := true
-			rs := &appsv1.ReplicaSet{
-				ObjectMeta: metav1.ObjectMeta{
-					Namespace: "default",
-					Name:      "rs-name",
-					Labels:    dep.Spec.Selector.MatchLabels,
-					OwnerReferences: []metav1.OwnerReference{
+			bldr := ControllerManagedBy(m).
+				For(&appsv1.Deployment{}).
+				Watches( // Equivalent of Owns
+					&source.Kind{Type: &appsv1.ReplicaSet{}},
+					&handler.EnqueueRequestForOwner{OwnerType: &appsv1.Deployment{}, IsController: true})
```

Comment:
Looking at this example, builder for controller appears to be a very *thin* improvement over current `controller` abstraction because users have to understand `manager`, `watches` and `handlers` to use these. Am I missing something ?

Reply:
this is just an example of using watches directly.  We need to test the "I want .Owns and one more complicated thing" -- the test just happens to use the Owner enqueuer for convienience